spec/controllers/lists_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+
+  before(:each) do
+    allow_message_expectations_on_nil
+  end
+
+  describe "GET #new" do
+    context "when logged in" do
+
+      before do
+        allow(controller).to receive(:current_user) { jack }
+      end
+
+      it "renders the new lists page" do
+        get :new
+        expect(response).to render_template("new")
+      end
+    end
+
+    context "when logged out" do
+      before do
+        allow(controller).to receive(:current_user) { nil }
+      end
+
+      it "redirects to the login page" do
+        get :new
+        expect(response).to redirect_to(new_session_url)
+      end
+    end
+  end
+
+  describe "POST #create" do
+
+    before do
+      allow(controller).to receive(:current_user) { jack }
+    end
+    context "when logged in" do
+
+      context "with invalid params" do
+        it "validates the presence of title and description" do
+          post :create, params: {list: { description: "this is an invalid list"} }
+          expect(response).to render_template("new")
+          expect(flash[:errors]).to be_present
+        end
+      end
+
+      context "with valid params" do
+        it "redirects to the  list's show page" do
+          post :create, params: {list: { name: "Things to Do", description: "do it"} }
+          expect(response).to redirect_to(list_url(List.last))
+        end
+      end
+    end
+
+    context "when logged out" do
+      before do
+        allow(controller).to receive(:current_user) { nil }
+      end
+
+      it "redirects to the login page" do
+        post :create, params: { list: {name: "Things to Do", description: "do it"} }
+        expect(response).to redirect_to(new_session_url)
+      end
+    end
+  end
+
+  describe "GET #index" do
+    context "when logged in" do
+
+      before do
+        allow(controller).to receive(:current_user) { jack }
+      end
+
+      it "renders the lists index" do
+        get :index
+        expect(response).to render_template("index")
+      end
+    end
+
+    context "when logged out" do
+      before do
+        allow(controller).to receive(:current_user) { nil }
+      end
+
+      it "redirects to the login page" do
+        get :index
+        expect(response).to redirect_to(new_session_url)
+      end
+    end
+  end
+
+  xdescribe "GET #show" do
+    create_jill_with_list
+
+    context "when logged in as the  list's owner" do
+
+      before do
+        allow(controller).to receive(:current_user) { jill }
+      end
+
+      it "renders the show page" do
+        get :show, params: {id: jill_list.id}
+        expect(response).to render_template(:show)
+      end
+    end
+
+    context "when logged in as a different user" do
+      before do
+        allow(controller).to receive(:current_user) { jack }
+      end
+
+      it "does not render the show page" do
+        begin
+          get :show, params: {id: jill_list.id}
+        rescue ActiveRecord::RecordNotFound
```

Comment:
[question] Is this getting hit here? What does this request render?

Reply:
The test here is supposed to simulate when a user that is not the owner of the list tries to see the show page of the given list. Prior to this line there is a part where the user "jack" gets signed in, the `get :show, params: {id: jill_list.id}` would be sending a get request to try and render `/lists/#{jill_list.id}`, and since jack does not own this list this action should not be allowed. The rescue is there just in case a redirect or other action hadn't yet been defined.